### PR TITLE
New version of callipyge

### PR DESCRIPTION
**Pull request by dinosaure**

### CODE REVIEW

**Comment by cfcs** [comment]
The test suite runs without error \o/

1) `_of_string` returns an exception, which doesn't seem to be documented:  `Exception: Invalid_argument "secret_key_of_string: invalid key"` - in this case the only check that is done is that the key has length `= 32`, so I think it would be nice if the error message reflected this invariant (like `"secret_key_of_string: key should consist of 32 bytes"`) ?

2) Public keys consisting of `let null = Callipyge.public_key_of_string (String.make 32 '\x00')` are accepted; this always results in a shared secret also consisting of nulls, and has been the source of a few practical protocol attacks on ECDH-stuff. This is kind of a controversial-ish topic because on one side you have the cryptographers going "that's completely okay because it doesn't leak private key material" and on the other side you have the applications programmers going "this crypto really sucks because it makes it much harder to design sensible protocols". I'd argue for rejecting 0-keys, I can find some links to discussions if you are interested.

3) I wonder what I'm doing wrong, but I was trying to implement the test vector from here https://tools.ietf.org/html/draft-ietf-tls-curve25519-01#appendix-A.2 , but this doesn't really do anything except fail on all my asserts, anyway - here goes my naive attempt:
```ocaml
open Callipyge
(* public keys *)
let p_a = public_key_of_string "\x85\x20\xF0\x09\x89\x30\xA7\x54\x74\x8B\x7D\xDC\xB4\x3E\xF7\x5A\x0D\xBF\x3A\x0D\x26\x38\x1A\xF4\xEB\xA4\xA9\x8E\xAA\x9B\x4E\x6A";;
let p_b = public_key_of_string "\xDE\x9E\xDB\x7D\x7B\x7D\xC1\xB4\xD3\x5B\x61\xC2\xEC\xE4\x35\x37\x3F\x83\x43\xC8\x5B\x78\x67\x4D\xAD\xFC\x7E\x14\x6F\x88\x2B\x4F";;

(* secret keys *)
let s_a = secret_key_of_string "\x6A\x2C\xB9\x1D\xA5\xFB\x77\xB1\x2A\x99\xC0\xEB\x87\x2F\x4C\xDF\x45\x66\xB2\x51\x72\xC1\x16\x3C\x7D\xA5\x18\x73\x0A\x6D\x07\x70";;
let s_b = secret_key_of_string "\x6B\xE0\x88\xFF\x27\x8B\x2F\x1C\xFD\xB6\x18\x26\x29\xB1\x3B\x6F\xE6\x0E\x80\x83\x8B\x7F\xE1\x79\x4B\x8A\x4A\x62\x7E\x08\xAB\x58";;

(* assert pubkey consistency *)
assert (public_of_secret s_a = p_a) ; (* fails because public_of_secret is defined as "%identity" ? what is the reasoning behind this? *)
assert (public_of_secret s_b = p_b) ;

let ss_ab = Array.make 32 '3' ;;
ecdh p_a s_b ss_ab ;; (* I would intuitively expect it to write the shared secret to this char array? *)

let ss_ba = Array.make 32 '4' ;;
ecdh p_b s_a ss_ba ;;

assert (ss_ab = ss_ba) ;;
```

**Comment by emillon** [comment]
- kudos for the diffstat :+1: 
- same remark about the test vectors, I'm only getting zeroes. I wrote a test suite [here](https://github.com/emillon/ocaml-noise/blob/2b46c4ca8e65761d2f8f3a7c19573e753f2a6ef1/test/unit/test_dh_25519.ml) if that helps (at the moment it's backed by tweetnacl)
- terminology: usually "secret key" is used for symmetric algorithms, and for asymmetric algorithms it's "public" and "private" (which is unfortunately a keyword)
- interface: the phantom types help but what do you think about naming the arguments? At least, if the array is to hold the result, it should be named `~into` or something. I'd also argue in favor of a returning an option or a result rather than raising an exception, but that's a style choice.

**Comment by dinosaure** [comment]
I got the problem but I need a time to fix it.

**Comment by cfcs** [comment]
@emillon Agreed, but would like to point out that `secret`/`private` is a `djb`-ism, like the `forward secrecy`/`key erasure` thing. In [the paper](https://cr.yp.to/ecdh/curve25519-20060209.pdf) it's consistently called secret key, and same in NaCl, so I think we should keep the name here, too.

**Comment by dinosaure** [comment]
Ok, I just look up precisely RFC and NaCl paper and inputs differ. I changed the intf. to be more close with what NaCl explains and this code works:

```ocaml
open Callipyge

let pp_unsafe ppf arr =
  for i = 0 to Array.length arr - 1 do Fmt.pf ppf "%02X" (Array.get arr i) done ;;

let s_a = secret_key_of_string "\x77\x07\x6d\x0a\x73\x18\xa5\x7d\x3c\x16\xc1\x72\x51\xb2\x66\x45\xdf\x4c\x2f\x87\xeb\xc0\x99\x2a\xb1\x77\xfb\xa5\x1d\xb9\x2c\x2a" ;;
let s_b = secret_key_of_string "\x5d\xab\x08\x7e\x62\x4a\x8a\x4b\x79\xe1\x7f\x8b\x83\x80\x0e\xe6\x6f\x3b\xb1\x29\x26\x18\xb6\xfd\x1c\x2f\x8b\x27\xff\x88\xe0\xeb" ;;

let p_a = public_of_secret s_a ;;
let p_b = public_of_secret s_b ;;

let pp_a = public_key_of_string "\x85\x20\xf0\x09\x89\x30\xa7\x54\x74\x8b\x7d\xdc\xb4\x3e\xf7\x5a\x0d\xbf\x3a\x0d\x26\x38\x1a\xf4\xeb\xa4\xa9\x8e\xaa\x9b\x4e\x6a" ;;
let pp_b = public_key_of_string "\xde\x9e\xdb\x7d\x7b\x7d\xc1\xb4\xd3\x5b\x61\xc2\xec\xe4\x35\x37\x3f\x83\x43\xc8\x5b\x78\x67\x4d\xad\xfc\x7e\x14\x6f\x88\x2b\x4f" ;;

let s_ab = shared s_a p_b ;;
let s_ba = shared s_b p_a ;;

Fmt.pr "public_of_secret(%a): %a.\n%!" pp_unsafe (s_a :> int array) pp_key p_a ;;
Fmt.pr "public_of_secret(%a): %a.\n%!" pp_unsafe (s_b :> int array) pp_key p_b ;;

Fmt.pr "shared(%a, %a): %a.\n%!" pp_unsafe (s_a :> int array) pp_key p_b pp_key s_ab ;;
Fmt.pr "shared(%a, %a): %a.\n%!" pp_unsafe (s_b :> int array) pp_key p_a pp_key s_ba ;;

assert (equal_key p_a pp_a) ;;
assert (equal_key p_b pp_b) ;;
assert (equal_key s_ab s_ba) ;;
```

**Comment by dinosaure** [comment]
Then, I will update `callipyge` to avoid `array` allocations and I think, we are ready for a new release.

**Comment by emillon** [comment]
This passes noise test vectors :+1:

**Comment by cfcs** [comment]
@dinosaure what do you think about rejecting the `null` case?
This still has the property that an attacker can derive the shared key without knowing the secret key corresponding to the `null` pk:
```
utop # let null = Callipyge.public_key_of_string (String.make 32 '\x00');;
val null : public key =
  [|0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 0;
    0; 0; 0; 0; 0; 0; 0|]
utop # shared s_a null;;
- : public key =
[|237; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255;
  255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255;
  255; 127|]
utop # shared s_a null;;
- : public key =
[|237; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255;
  255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255; 255;
  255; 127|]
```

**Comment by dinosaure** [comment]
I don't have a strong opinion about that and raise an exception in this case is not a problem. However I don't know what I need to say to the client 😄 !

**Comment by cfcs** [comment]
@dinosaure could just be "null public key" or something like that?

**Comment by dinosaure** [comment]
Done 👍 we just need a documentation now and finally release it.

About allocation, computation of `curve25519` allocates only one `array` now. We need to use `ctx` to do computation now and avoid pressure on GC.

**Comment by cfcs** [comment]
BTW, `val Callipyge.shared secret key -> public key -> public key` - it's a bit confusing that it returns a `public_key` when the shared value is not a public key?

**Comment by emillon** [comment]
In terms of types, this is `shared : scalar -> point -> point`, so it's not too much a problem if the result is typed `public_key`, but it's true that a `shared` type is cleaner.

File: lib/callipyge.ml [hunk 1]
```diff
@@ -4,260 +4,132 @@ let basev =
...
+  u := !u + (Array.get a (a_offset + 31));
+  Array.set a (a_offset + 31) !u
 
-  (* XXX(dinosaure): Allocation here! *)
+let freeze a a_offset =
+  let a_orig = Array.make 32 0 in
 
   for j = 0 to 31 do
-    A.set a_orig j (A.get a (a_offset + j));
+    Array.set a_orig j (Array.get a (a_offset + j));
```
Comment (cfcs) [inline lib/callipyge.ml]: How about
```ocaml
let a_orig = Array.init 32 (fun j -> Array.get a (a_offset +j) ) in
```
here?

**Comment by dinosaure** [comment]
At this stage, we have something which work and performance is not so bad. I think, it's ready for a release if people are happy with API.  We can continue to improve performance with some tricks but it's about the next release.

Some questions stills open like the `eq` function which come from `digestif` (see mirage/digestif#50).

**Comment by dinosaure** [comment]
Ah and the documentation obviously ...

**Comment by dinosaure** [comment]
Ok I believe you 👍 .

**Comment by dinosaure** [comment]
See [@dinosaure/eqaf](https://github.com/dinosaure/eqaf)

**Pull request merged**